### DISCUSSION TITLE
Generate wasm

### DISCUSSION THREAD

**Discussion by kindofwhat** [chaosprint/glicol#119] [Q&A]
I am trying to generate the wasm package and use it in the simple web application.

for this

- I commented out the `windows.verison`line in`js\src\glicol.js`
- I executed the `build.sh` in `rs\wasm`which
- I started the web app with `npm run dev` in `js`

This lead to following exception in the browser:

```
Import #0 module="__wbindgen_placeholder__" error: module is not an object or function
```

With my absolute minimal knowledge about the wasm ecosystem, it looks like there might be some missing wbindgen config, but I am at a loss where this is missing.

Is there another way to build glicol for the web?

**Comment by chaosprint** [comment]
Are you trying to use it with other tools such as https://github.com/rustwasm/wasm-bindgen?

You can try to build manually:
```
cargo build --target wasm32-unknown-unknown --release
```

Then you can find the `.wasm` file in `./rs/wasm/target/wasm32-unknown-unknown/release/glicol_wasm.wasm`.

Copy that one to your project and load the `wasm` file from JS.

**Reply by kindofwhat** [reply]
Hi @chaosprint thank you for your answer
That is exactly what the build script does and it leads to above mentioned.

I wonder if `wasm-bindgen`is somehow "polluting" my build chain, but beats me where this could happen...

**Reply by kindofwhat** [reply]
ok, I am kind of one step further, or two steps behind, dependes on the pov

I added following line to `dependencies` of `Cargo.toml`:

```
getrandom = { version = "0.2", features = ["js"] }
```

this loads the bindgen as dependency.

Now, when I try to build without this line I get the following error message when building:

```
   Compiling getrandom v0.2.7
error: the wasm32-unknown-unknown target is not supported by default, you may need to enable the "js" feature. For more information see: https://docs.rs/getrandom/#webassembly-support
   --> /Users/.../.cargo/registry/src/github.com-1ecc6299db9ec823/getrandom-0.2.7/src/lib.rs:235:9
    |
235 | /         compile_error!("the wasm32-unknown-unknown target is not supported by \
236 | |                         default, you may need to enable the \"js\" feature. \
237 | |                         For more information see: \
238 | |                         https://docs.rs/getrandom/#webassembly-support");
    | |________________________________________________________________________^

```
So the question is currently how to build it correctly with js support but without dragging bindgen into it?

**Reply by chaosprint** [reply]
Yes I know this issue from `getrandom`.

But I think it's better to ask `wasm-bindgen` community for the solution if you need to use that framework. 

You can view Glicol as a black box that takes code in and calculate audio blocks out so in theory it should be fine to work with wasm-bindgen.

But the way to get the best audio performance from browser is what's shown in the `js` folder, and I wrote quite some JS code rather than using `wasm-bidgen`.

I guess for the latter, you write most of your logic in Rust rather than JS.

Another way is to use https://glicol.js.org/

I haven't updated it for a while but you can check the examples there to see if it fits your project.
